File: Lab07-NumericalDifferentiation/lab7.py
def oblicz_tablice_roznic_skonczonych(y):
    n = len(y)
    tabela = [[0.0] * n for _ in range(n)]
    for i in range(n):
        tabela[i][0] = y[i]
    for j in range(1, n):
        for i in range(n - j):
            tabela[i][j] = tabela[i + 1][j - 1] - tabela[i][j - 1]
    return tabela

def pochodna_taylor_przod(tabela, h, n):
    wynik = 0.0
    for k in range(1, n):
        term = tabela[0][k]
        znak = 1.0 if k % 2 != 0 else -1.0
        wynik += znak * (1.0 / k) * term
    return wynik / h

def pochodna_taylor_wstecz(tabela, h, n):
    wynik = 0.0
    for k in range(1, n):
        wynik += (1.0 / k) * tabela[n - 1 - k][k]
    return wynik / h

def pochodna_stirling(tabela, h, idx):
    n = len(tabela)
    avg = (tabela[idx][1] + tabela[idx - 1][1]) / 2.0

    return avg / h

def roznice_dzielone(x, y):
    n = len(x)
    dd = [[0.0] * n for _ in range(n)]
    for i in range(n):
        dd[i][0] = y[i]
    for j in range(1, n):
        for i in range(n - j):
            dd[i][j] = (dd[i + 1][j - 1] - dd[i][j - 1]) / (x[i + j] - x[i])
    return [dd[0][i] for i in range(n)]

def wypisz_wielomian(x, a):
    txt = f"W(x) = {a[0]:.4f}"
    for k in range(1, len(a)):
        znak = " + " if a[k] >= 0 else " - "
        skladnik = f"{abs(a[k]):.4f}"
        for i in range(k):
            skladnik += f"(x - {x[i]})"
        txt += znak + skladnik
    return txt

n = int(input("Podaj liczbę punktów n: "))
x = []
y = []
print("Wpisz x:")
for i in range(n): x.append(float(input(f"x[{i}] = ")))
print("Wpisz y:")
for i in range(n): y.append(float(input(f"y[{i}] = ")))

h = x[1] - x[0]
tab_roznic = oblicz_tablice_roznic_skonczonych(y)

while True:
    txt_in = input("Podaj x dla pochodnej (lub Enter aby wyjść): ")
    if not txt_in: break
    x_pt = float(txt_in)
    
    idx = -1
    for i in range(n):
        if abs(x[i] - x_pt) < 1e-9:
            idx = i
            break
            
    if idx == -1:
        print("Punkt musi być węzłem")
        continue
        
    if idx == 0:
        val = pochodna_taylor_przod(tab_roznic, h, n)
        metoda = "Taylor (Progresywna)"
    elif idx == n - 1:
        val = pochodna_taylor_wstecz(tab_roznic, h, n)
        metoda = "Taylor (Wsteczna)"
    else:
        val = pochodna_stirling(tab_roznic, h, idx)
        metoda = "Stirling (Centralna)"
        
    print(f"Metoda: {metoda}")
    print(f"f'({x_pt}) = {val:.5f}")

print("\nWielomian interpolacyjny Newtona:")
a_coeffs = roznice_dzielone(x, y)
print(wypisz_wielomian(x, a_coeffs))

File: Lab07-NumericalDifferentiation/lab7.m
function lab7matlab()
    clc;
    n = input('Podaj liczbe punktow n: ');
    xi = zeros(1,n);
    yi = zeros(1,n);

    disp('Wpisz x:');
    for i=1:n
        xi(i) = input(sprintf('x(%d) = ', i));
    end
    disp('Wpisz y:');
    for i=1:n
        yi(i) = input(sprintf('y(%d) = ', i));
    end

    h = xi(2) - xi(1);
    D = zeros(n, n);
    D(:,1) = yi';
    for j=2:n
        for i=1:(n-j+1)
            D(i,j) = D(i+1,j-1) - D(i,j-1);
        end
    end

    x_pt = input('Podaj x dla pochodnej: ');
    if ~isempty(x_pt)
        idx = find(abs(xi - x_pt) < 1e-9);
        if isempty(idx)
            disp('Punkt musi byc wezlem.');
        else
            if idx == 1
                der = taylor_przod(D, h, n);
                met = 'Taylor (Progresywna)';
            elseif idx == n
                der = taylor_tyl(D, h, n);
                met = 'Taylor (Wsteczna)';
            else
                der = stirling(D, h, idx, n);
                met = 'Stirling (Centralna)';
            end
            fprintf('Metoda: %s\n', met);
            fprintf('f''(%g) = %.5f\n', x_pt, der);
        end
    end

    dd = zeros(n,n);
    dd(:,1) = yi(:);
    for j=2:n
        for i=1:(n-j+1)
            dd(i,j) = (dd(i+1,j-1) - dd(i,j-1)) / (xi(i+j-1) - xi(i));
        end
    end
    a = dd(1,:);
    fprintf('\nW(x) = %.4f', a(1));
    for k=2:n
        if a(k) >= 0, s = '+'; else, s = '-'; end
        fprintf(' %s %.4f', s, abs(a(k)));
        for j=1:(k-1)
            fprintf('(x-%.4g)', xi(j));
        end
    end
    fprintf('\n');
end

function res = taylor_przod(D, h, n)
    res = 0;
    for k=1:(n-1)
        res = res + ((-1)^(k-1)) * (1/k) * D(1, k+1);
    end
    res = res / h;
end

function res = taylor_tyl(D, h, n)
    res = 0;
    for k=1:(n-1)
        res = res + (1/k) * D(n-k, k+1);
    end
    res = res / h;
end

function res = stirling(D, h, idx, n)
    avg = (D(idx, 2) + D(idx-1, 2)) / 2;
    res = avg / h;
end